packages/core/components/DirectoryTree/findChildNodes.ts

```diff
@@ -0,0 +1,113 @@
+import { defaults, isEmpty, pull } from "lodash";
+
+import { NO_VALUE_NODE, ROOT_NODE } from "./directory-hierarchy-state";
+import FileFilter, { FilterType } from "../../entity/FileFilter";
+import FileSet from "../../entity/FileSet";
+import { AnnotationService } from "../../services";
+import { naturalComparator } from "../../util/strings";
+
+export interface FindChildNodesParams {
+    ancestorNodes?: string[];
+    currentNode: string;
+    hierarchy: string[];
+    annotationService: AnnotationService;
+    fileSet?: FileSet;
+    selectedFileFilters?: FileFilter[];
+    shouldShowNullGroups?: boolean;
+}
+
+const DEFAULTS = {
+    ancestorNodes: [],
+    currentNode: ROOT_NODE,
+    selectedFileFilters: [],
+    shouldShowNullGroups: false,
+};
+
+export async function findChildNodes(params: FindChildNodesParams): Promise<string[]> {
+    const {
+        ancestorNodes,
+        currentNode,
+        fileSet,
+        hierarchy,
+        annotationService,
+        selectedFileFilters,
+        shouldShowNullGroups,
+    } = defaults({}, params, DEFAULTS);
+
+    const isRoot = currentNode === ROOT_NODE;
+
+    const combinedFileFilters = selectedFileFilters.concat(
+        (fileSet?.filters || [])?.filter((filter) => !selectedFileFilters.includes(filter))
+    );
+
+    // If at root of hierarchy, currentNode will be set to "ROOT_NODE"
+    // We trim that from the path as it is not meaningful in this context
+    const pathToNode = pull([...ancestorNodes, currentNode], ROOT_NODE);
+
+    let values: string[] = [];
+
+    const depth = pathToNode.length;
+    const annotationNameAtDepth = hierarchy[depth];
+    const userSelectedFiltersForCurrentAnnotation = combinedFileFilters
+        .filter(
+            (filter) => filter.name === annotationNameAtDepth && filter.type !== FilterType.EXCLUDE
+        )
+        .map((filter) => filter.value);
+
+    if (isRoot) {
+        const rootHierarchy = shouldShowNullGroups ? [annotationNameAtDepth] : hierarchy;
```

Comment:
Is `hierarchy` an empty list here? Could you add a comment explaining a bit of the difference between when we want `null` nodes and don't here?

Reply:
Will add! Hierarchy is not an empty list, it's the whole group structure. When we want `null` nodes, we should ignore the rest of the hierarchy and only look at the values for the current annotation. Using the hierarchy implicitly adds `include` filters for all the other fields, which forces us to drop any values at the current level where something down the line is missing a value